Fixed multithreaded gradle

malice00 opened the pull request:
This PR fixes #1335 , although the solution for that issue may need to be redone if someone can reproduce the actual use-case -- instead of a non-existent parameter, a hard-coded String is now used.

After fixing that, I found a several of other problems with the multithreaded solution:
- Arguments specific for a task in the wrong place
- Arguments specific for a task must be added for every call to the task
- Unlike in the single runs, gradle fails all jobs if just one task has an issue and NO dependencies are returned

So, the necessary arguments have been multiplied, all arguments have been arranged to their correct places and a new envvar has been introduced to be able to skip problematic modules.

prabhu commented:
@ajmalab could you advise and work with @malice00 to get this ready?

prabhu commented on utils.js:
```diff
@@ -2878,40 +2878,37 @@ export function parseGradleProperties(rawOutput) {
 export function executeParallelGradleProperties(dir, rootPath, allProjectsStr) {
   const defaultProps = {
-    rootProject: subProject,
+    rootProject: "root",
```
Instead of hardcoding, can we set this to `spstr` from the for loop?

malice00 replied:
I'll have look, this was just a quick fix/hack to get things working.

malice00 commented:
I've renamed some of the variables to make the code less confusing. Some variables where still called 'purl' (or some variant thereof) but where actually a 'bom-ref'.
Also rebased to the current master.

prabhu commented on docs/ENV.md:
```diff
@@ -24,6 +24,7 @@ The following environment variables are available to configure the bom generatio
 | GRADLE_HOME                  | Specify gradle home                                                                                                                                                                                                                                                                                              |
 | GRADLE_CMD                   | Set to override gradle command                                                                                                                                                                                                                                                                                   |
 | GRADLE_DEPENDENCY_TASK       | By default cdxgen use the task "dependencies" to collect packages. Set to override the task name.                                                                                                                                                                                                                |
+| GRADLE_SKIP_MODULES          | Comma-separated list of modules to skip during the "dependencies" task. This can be useful if you have modules that would fail the gradle build, eg when they do not have dependencies in the given configuration. Use "root" if the top most module should be skipped, use the name (without leading ":") for all others. |
```
Can we add an example to the repo tests?

malice00 commented:
I completely removed the object, since this method should return a String -- even according to its documentation.

prabhu commented:
Interesting. Can we add some repo tests to prove that the sbom generated with and without multi threading is identical? We can use this [project](https://github.com/AppThreat/custom-json-diff) similar to our snapshot tests.

malice00 commented:
I'll take a stab at it...

prabhu commented:
Please let me know once you are done and ready.

malice00 commented:
Honestly, I'm unsure on how to get a good test in here without setting up an actual (small) project, which would then need Gradle (and therefore Java) and probably some other stuff... Maybe it will be easier to do something like this when you redesign cdxgen in the next major version...
But if anyone has any ideas on how to get a test up and running that can generate the whole SBOM, I'm open to suggestions!

prabhu commented:
Java and gradle are already setup in repotests, so we only need a sample repo.

malice00 commented:
Ah, great! I'm not seeing them though -- where are these tests located, so I can check out what we'd need and how to get it running?

prabhu commented:
https://github.com/CycloneDX/cdxgen/blob/master/.github/workflows/repotests.yml

malice00 commented:
I added a test, but for some reason it's not doing anything... Even with `CDXGEN_DEBUG_MODE` I don't see anything -- how do I get more output from cdxgen to figure out what might be wrong?

As for that cjd project, I added it and a call to it, but I see no way how this may help in a test -- it simply generates a report which must manually be checked, or am I missing something?

prabhu commented:
@aryan-rajoria could you kindly take a look and suggest some repo tests? We can merge this PR and add the tests separately if it might be easier.

malice00 commented:
Does anybody know why we don't see the output of cdxgen in these repotests? Would make stuff a lot easier to check/debug...
Also, wouldn't it be better if the 'tests' actually had assertions or something like that? Right now they seem to just run a build and only fail when something goes horribly wrong. My tests should eg fail if (like now) the SBOM has no components. And a check like you suggested with cjb would be great, if only that tool could actually fail the build, eg with an exit code if the SBOMs are different...
But maybe this is better in its own issue though. 😄

prabhu commented:
@malice00 good points raised. We tried coming up with snapshot tests to track the diff and fail the PR. Testing, in general, is going to take some more effort to get there. As part of another PR, I have temporarily hardcoded "root" as the subProject name. Since there is no response from Ajmal, can we park this PR and resume once the refactoring as part of #1360 is complete?

malice00 commented:
@prabhu, I'm not thrilled about that, because I'm working on some other changes that build on this. Any idea when #1360 will be merged?

prabhu commented:
@setchy are you happy to get this merged and ported to the refactor branch? Will be great if we all could collaborate since freezing the project for a large scale refactor is proving difficult

setchy commented:
No need to freeze the project - happy to merging ongoing updates into the refactor branch. I don't want that initial refactor to be more than just moving files around and import updates

malice00 commented:
@prabhu Sure, no problem.
